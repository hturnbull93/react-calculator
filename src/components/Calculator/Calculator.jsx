import React, { Component } from "react";
import "./Calculator.css";
import Display from "../Display/Display";
import Keypad from "../Keypad/Keypad";

class Calculator extends Component {
  state = {
    displayValue: "0",
    numbers: ["9", "8", "7", "6", "5", "4", "3", "2", "1", ".", "0", "ce"],
    operators: ["/", "x", "-", "+"],
    selectedOperator: "",
    storedValue: "",
  };

  callOperator = () => {};

  setOperator = () => {};

  updateDisplay = (value) => {
    let { displayValue } = this.state;
    if (value === "." && displayValue.includes(".")) return

    let newDisplayValue
    if (displayValue === "0") {
      displayValue = "";
    }
    newDisplayValue = value === "ce" ? displayValue.slice(0, -1) : displayValue.concat(value)

    if (newDisplayValue === "") {
      newDisplayValue = "0"
    }
    this.setState({ displayValue: newDisplayValue });
  };

  render() {
    const { displayValue, numbers, operators } = this.state;
    return (
      <div className="calculator-container">
        <Display displayValue={displayValue} />
        <Keypad
          callOperator={this.callOperator}
          numbers={numbers}
          operators={operators}
          setOperator={this.setOperator}
          updateDisplay={this.updateDisplay}
        />
      </div>
    );
  }
}

export default Calculator;
