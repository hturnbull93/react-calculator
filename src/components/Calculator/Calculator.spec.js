import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import Calculator from "./Calculator";
import Display from "../Display/Display";
import Keypad from "../Keypad/Keypad";

describe("Calculator", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a div", () => {
    const wrapper = shallow(<Calculator />);
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("renders the Display Component", () => {
    const wrapper = shallow(<Calculator />);
    const display = (
      <Display displayValue={wrapper.instance().state.displayValue} />
    );
    expect(wrapper.containsMatchingElement(display)).toEqual(true);
  });

  it("renders the Keypad Component", () => {
    const wrapper = shallow(<Calculator />);
    const keypad = (
      <Keypad
        callOperator={wrapper.instance().callOperator}
        numbers={wrapper.instance().state.numbers}
        operators={wrapper.instance().state.operators}
        setOperator={wrapper.instance().setOperator}
        updateDisplay={wrapper.instance().updateDisplay}
      />
    );
    expect(wrapper.containsMatchingElement(keypad)).toEqual(true);
  });

  describe("Functionality", () => {
    it("calls updateDisplay when a number key is clicked", () => {
      const wrapper = mount(<Calculator />);
      const spy = jest.spyOn(wrapper.instance(), "updateDisplay");
      wrapper.instance().forceUpdate();
      expect(spy).toHaveBeenCalledTimes(0);
      wrapper.find(".number-key").first().simulate("click");
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("calls setOperator when an operator key is clicked", () => {
      const wrapper = mount(<Calculator />);
      const spy = jest.spyOn(wrapper.instance(), "setOperator");
      wrapper.instance().forceUpdate();
      expect(spy).toHaveBeenCalledTimes(0);
      wrapper.find(".operator-key").first().simulate("click");
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("calls callOperator when the submit key is clicked", () => {
      const wrapper = mount(<Calculator />);
      const spy = jest.spyOn(wrapper.instance(), "callOperator");
      wrapper.instance().forceUpdate();
      expect(spy).toHaveBeenCalledTimes(0);
      wrapper.find(".submit-key").first().simulate("click");
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe("updateDisplay", () => {
    it("updates displayValue", () => {
      const wrapper = shallow(<Calculator />);
      wrapper.instance().updateDisplay("5");
      expect(wrapper.state("displayValue")).toEqual("5");
    });
    
    it("concatenates displayValue", () => {
      const wrapper = shallow(<Calculator />);
      wrapper.instance().updateDisplay("5");
      wrapper.instance().updateDisplay("0");
      expect(wrapper.state("displayValue")).toEqual("50");
    });
  });
});
