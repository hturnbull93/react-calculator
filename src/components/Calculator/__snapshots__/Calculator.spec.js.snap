// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  className="calculator-container"
>
  <div
    className="display-container"
  >
    <p
      className="display-value"
    >
      0
    </p>
  </div>
  <div
    className="keypad-container"
  >
    <div
      className="numbers-container"
    >
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          9
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          8
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          7
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          6
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          5
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          4
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          3
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          2
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          1
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          .
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          0
        </p>
      </div>
      <div
        className="key-container number-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          ce
        </p>
      </div>
    </div>
    <div
      className="operators-container"
    >
      <div
        className="key-container operator-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          /
        </p>
      </div>
      <div
        className="key-container operator-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          x
        </p>
      </div>
      <div
        className="key-container operator-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          -
        </p>
      </div>
      <div
        className="key-container operator-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          +
        </p>
      </div>
    </div>
    <div
      className="submit-container"
    >
      <div
        className="key-container submit-key"
        onClick={[Function]}
      >
        <p
          className="key-value"
        >
          =
        </p>
      </div>
    </div>
  </div>
</div>
`;
